File: 2022_11_03_leetcode_in_fsharp/LeetCode/LeetCode/P1342_NumberOfSteps.fs
// https://leetcode.com/problems/number-of-steps-to-reduce-a-number-to-zero/

module P1342_NumberOfSteps

let numberOfSteps num =
    2 * num // +1 to the final result
    |> Seq.unfold (fun n ->
        let x = if n % 2 = 0 then n / 2 else n - 1
        Some(x, x))
    |> Seq.takeWhile (fun n -> n <> 0)
    |> Seq.length


let _ = numberOfSteps 14 // -> 6
let _ = numberOfSteps 8 // -> 4
let _ = numberOfSteps 123 // -> 12
let _ = numberOfSteps 8 // -> 4
let _ = numberOfSteps 123 // -> 12

File: 2022_11_19_leetcode_in_csharp/LeetCode/P1342_NumberOfSteps.cs
// https://leetcode.com/problems/number-of-steps-to-reduce-a-number-to-zero/

namespace LeetCode;

using static System.Linq.Enumerable;

class P1342_NumberOfSteps
{
    static int NumberOfSteps(int num)
    {
        return new[] { num }
            .Expand(n => new[] { (n % 2 == 0 ? n / 2 : n - 1) })
            .TakeWhile(n => n != 0)
            .Count();
    }

    public static void Run()
    {
        var inputs = new[] { 14, 8, 123, 8, 123 };

        foreach (var input in inputs)
        {
            var output = NumberOfSteps(input);
            Console.WriteLine($" {input} -> {output}");
        }
    }
}